### DISCUSSION TITLE
[Proposal] Have 'is' operator type pattern assign value if right-hand variable already exists within scope

### DISCUSSION THREAD

**Discussion by bobbymonboy** [dotnet/csharplang#8814] [Language Ideas]
For the entirety of this post, I will refer to the **is** operator in the context of the specific type pattern `item1 is T item2`.

Currently, if an **is** operator's right-hand variable has the same name as one that already exists in the scope, a compiler error occurs. I believe it may simplify some code if a change were made:
- If the **is** operator returns `true`, it casts `item1` as type `T` to `item2`.
- If the **is** operator returns `false`, it sets `item2` to the default for its type.

This would allow you to shorten something like this:
```
public bool ConvertChild<TDerived>(Container container, out TDerived child) where TDerived : Container
{
	if (container is TDerived containerChild)
	{
		child = containerChild;
		return true;
	}
	else
	{
		child = default;
		return false;
	}
}
```
To this:
```
public bool ConvertChild<TDerived>(Container container, out TDerived child) where TDerived : Container => container is TDerived child;
```
However, I do see an issue with this. If **is** assigns something to an already-existing variable whether or not it fails, it practically becomes an assignment expression. Therefore, **is** may need to be considered a statement if this change is implemented.

Very importantly, because currently `item2` will be unassigned outside of whatever **if** statement the **is** operator is used in, any already existing code that uses `item2` outside of the **if** statement will have already have been required to assign a value to it before using it. Since all pre-existing code will already have assigned a value to `item2` before this change, giving it a default value outside the **if** statement won't affect anything, as it will have already been overridden. Therefore, this change shouldn't break any currently written code for anybody.

**Comment by theunrepentantgeek** [comment]
This would essentially allow variable aliasing to happen through the back door - and the LDM have historically been very resistant to variable aliasing due to the number of subtle bugs it introduces.

It's fairly regular that this comes up, but with a quick search I only found #1741; others may be able to provide references to other issues.

**Comment by huoyaoyuan** [comment]
> To this:
> 
> ```
> public bool ConvertChild<TDerived>(Container container, out TDerived child) where TDerived : Container => container is TDerived child;
> ```

It seems that you forget the `as` operator:
```csharp
public bool ConvertChild<TDerived>(Container container, out TDerived child) where TDerived : Container
    => child = container as TDerived;
```

**Comment by canton7** [comment]
@huoyaoyuan That won't compile -- the method should return a `bool`

**Comment by john-h-k** [comment]
```cs
public bool ConvertChild<TDerived>(Container container, out TDerived child) where TDerived : Container
    => (child = container as TDerived) is object;
```

not very pretty tho
